File: StringToInteger/string_to_integer.py
class Solution:
    def myAtoi(self, s: str) -> int:

        s = s.strip() 
        if not s:
            return 0

        sign = 1
        index = 0
        val = 0

        if s[0] == "-":
            sign = -1
            index += 1
        elif s[0] == "+":
            index += 1

        while index < len(s) and s[index].isdigit():
            val = val * 10 + int(s[index])

            if val > 2147483647 or val < -2147483648:
                if sign == 1:
                    return 2147483647
                else:
                    return -2147483648

            index += 1

        return sign * val

File: StringToInteger/string_to_integer.go
package main

func myAtoi(s string) int {
	if len(s) == 0 {
		return 0
	}

	sign := 1 // -1 negative
	index := 0
	val := 0

	for len(s) > index && s[index] == ' ' {
		index++
	}

	if index >= len(s) {
		return 0
	}

	if v := s[index]; v == '-' || v == '+' {
		if v == '-' {
			sign = -1
		}

		index++
	}

	for {
		if index == len(s) {
			break
		}

		if isNumericRune(s[index]) == false {
			return sign * val
		}

		val = val*10 + int(s[index]) - '0'

		if val > 2147483647 || val < -2147483648 {
			if sign == 1 {
				return 2147483647
			}

			return -2147483648
		}

		index++
	}

	return sign * val
}

func isNumericRune(x byte) bool {
	return x >= '0' && x <= '9'
}
